Patterns: merge unsynced into inserter patterns tab and add paging and filtering

glendaviesnz opened the pull request:
## What?
Merges the synced patterns tab into the main patterns tab and adds an `All patterns` category and paging and filtering in the patterns inserter tab and patterns explorer.

## Why?
Currently it is confusing that synced patterns are in a separate tab, and also causes performance issues because patterns are not paged.

## How?

- Removes the synced patterns tab from inserter
- Updates the store patterns selector to get synced and unsynced patterns so both display in the existing patterns tab
- Adds paging to the patterns list to account for large numbers of synced patterns in some sites
- Adds a filter to allow users to limit pattern list to just theme, or user patterns, and another filter when user patterns selected to limit to synced or unsynced patterns

This PR was part of a group of PRs that were merged back into https://github.com/WordPress/gutenberg/pull/53835 before being merged into trunk. This PR was branched off https://github.com/WordPress/gutenberg/pull/53933.

A full list of the PRs involved in this work can be [found here](https://github.com/WordPress/gutenberg/issues/53164).

## Testing Instructions

- In the post editor add some synced and unsynced patterns with categories assigned
- Check that the patterns appear in the correct categories in the inserter patterns tab
- Check that the `All patterns` tab shows all patterns and that paging at the bottom of the tab works
- Check that the Source Filter select list works as expected and that correct patterns display for each selected filter
- Also check that when the `My patterns` source filter is selected that a `Sync type filter` appears at the top of the patterns list and works as expected
- Check that changing Source or Sync type filters when categories are selected resets the category to `All`
- Repeat all of the above in the patterns explorer modal
- With screenreader/voiceover running try selecting a pattern source from dropdown at top of patterns tab and check that the number of category buttons displayed is announced correctly

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/3629020/e377fe33-675e-4344-bdfd-e63141e74a3f

glendaviesnz committed:
Add sync status back

glendaviesnz committed:
Add back pattern id

glendaviesnz committed:
Add my patterns toggle

glendaviesnz committed:
Remove sync toggle for now

glendaviesnz committed:
merge synced patterns into patterns tab and add paging

glendaviesnz committed:
Make sort order same as site editor

glendaviesnz committed:
Fix bug with paging

glendaviesnz committed:
Reset to page one if search filter changes

glendaviesnz committed:
Update e2e tests

glendaviesnz committed:
rename snapshot file

glendaviesnz committed:
Add pattern type filters to insert patterns tab

glendaviesnz committed:
Add screen reader announcement to indicate number of categories when pattern filter changes

glendaviesnz committed:
Add aria label to filter select

glendaviesnz committed:
Add filtering to patterns explorer

glendaviesnz committed:
Simplify filter

glendaviesnz commented:
@alexstine, @joedolson, @afercia I have added a new filter into the post editor inserter tab to allow users to show all, or theme, or user synced or user unsynced patterns. Selecting a filter will then limit the category buttons shown to only those that contain patterns of that type.

I have used [a simple select list ](https://github.com/WordPress/gutenberg/pull/54007/files#diff-044e38d347b56956b985d170a094b7dea23d21314737ab99545f00025ea37857R19) for this, and also added [a screen reader announcement ](https://github.com/WordPress/gutenberg/pull/54007/files#diff-6987beb6cf61ef9eeebfc8c58c8352eb1a736edada51f11a169d9d4ec0a218c7R108) that gives the number of category buttons displayed when the filter is changed. Can you see any problems with this approach from an accessibility point of view?

There is another [suggested approach here](https://github.com/WordPress/gutenberg/issues/53033) to have an icon and label at the very bottom of the tab which opens a dropdown/modal with the filters to select. This approach seems less useful for keyboard users as you need to tab through all the current category buttons before encountering the filtering options. 

The icon and dropdown could go at the top of the tab, but a select list has the advantage of clearly indicating which filter is currently applied. 

FYI, this initial implementation for 6.4 won't have multiple filter selection as an option, it just allows one filter at a time to be applied.

glendaviesnz commented:
@dcalhoun, @geriux it would probably be good for someone from mobile to cast their eyes over this one. It merges the  `synced patterns/reusable blocks` tab into the `Patterns` tab and adds a filter to limit display to certain types. It looks native generates its own inserter tabs, so I don't think mobile will be affected by this change, but just double-checking. Thanks.

glendaviesnz committed:
Fix test snapshots

aaronrobertshaw commented on packages/block-editor/src/components/inserter/block-patterns-filter.js:
```diff
@@ -0,0 +1,31 @@
+				aria-label={ __( 'Filter patterns by type' ) }
+			/>
+		</>
+	);
```
```suggestion
	return (
		<SelectControl
			label={ __( 'Filters' ) }
			options={ SYNC_FILTERS }
			value={ value }
			onChange={ onChange }
			aria-label={ __( 'Filter patterns by type' ) }
		/>
	);
```

It doesn't appear that we need the fragment here.

glendaviesnz replied:
Fixed.

aaronrobertshaw commented on packages/block-editor/src/components/inserter/block-patterns-tab.js:
```diff
@@ -27,28 +28,38 @@ import usePatternsState from './hooks/use-patterns-state';
+						( filter === PATTERN_TYPES.synced &&
+							pattern.syncStatus === '' ) ||
+						( filter === PATTERN_TYPES.theme &&
+							! pattern.name.startsWith( 'core/block' ) )
```
We use this set of conditions a few times. Can we extract that at all aiding readability?

glendaviesnz replied:
Done

aaronrobertshaw commented on packages/block-editor/src/components/inserter/block-patterns-explorer/patterns-list.js:
```diff
@@ -62,21 +69,36 @@ function PatternList( { filterValue, selectedCategory, patternCategories } ) {
+					( filterValue === PATTERN_TYPES.synced &&
+						pattern.syncStatus !== '' ) ||
+					( filterValue === PATTERN_TYPES.unsynced &&
+						pattern.syncStatus !== PATTERN_TYPES.unsynced )
```
This set of conditions is reused a couple of times. Can we refactor that?

glendaviesnz replied:
Done.

aaronrobertshaw commented on packages/block-editor/src/components/inserter/block-patterns-explorer/patterns-list.js:
```diff
@@ -62,21 +69,36 @@ function PatternList( { filterValue, selectedCategory, patternCategories } ) {
+										category
+									)
+							)
+					: pattern.categories?.includes( selectedCategory );
```
Could we make this eaiser to grok as well? Maybe something like the following that continues using `if` instead of the ternary and doesn't need to check every pattern category?

```js
				if ( selectedCategory === 'uncategorized' ) {
					const hasKnownCategory = pattern.categories.some(
						( category ) =>
							registeredPatternCategories.includes( category )
					);

					return ! pattern.categories?.length || ! hasKnownCategory;
				}

				return pattern.categories?.includes( selectedCategory );
```

glendaviesnz replied:
Done.

aaronrobertshaw commented on packages/block-editor/src/components/inserter/block-patterns-tab.js:
```diff
@@ -165,10 +200,21 @@ export function BlockPatternsCategoryPanel( {
+		numPages,
+		changePage,
+		currentPage,
+	} = usePatternsPaging(
```
Nit: Could we reduce the code a bit here and return pagination props from the hook as an object and just spread them on the `BlockPatternsPaging` component? We have to plumb them through in a couple of places so could be a nice touch.

glendaviesnz replied:
Done.

aaronrobertshaw reviewed (commented):
Huge effort wrangling all the pattern category inserter updates @glendaviesnz 💪 

In general, this is testing really well and I'm looking forward to seeing pattern categories land in 6.4.

While testing I did run into a couple of minor issues:

1. When searching in the inserter you see the synced patterns twice in the results (once as an icon for blocks and a second time in the patterns)
2. The synced patterns icon looks a bit odd in the search results when it's combined with the pattern name.
    <img width="350" alt="Screenshot 2023-09-07 at 6 34 00 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/533dc856-efe5-4d66-b895-f9a98739c543">
3. We probably need to cover the "Show button text labels" preference for the pagination as well
    <img width="294" alt="Screenshot 2023-09-07 at 6 45 51 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/2c14430e-a0ed-4956-90d6-a5cf00af1a41">
4. When switching between filters, if the previously selected category isn't in the available categories after filtering, it appears like there are no patterns for the filter. Selecting one of the available categories does correctly show their patterns though.

https://github.com/WordPress/gutenberg/assets/60436221/8672ab44-f8ce-4517-b454-c26c4af89f46

Thanks again for working through this one!

kevin940726 commented on packages/block-editor/src/components/block-patterns-list/index.js:
```diff
@@ -101,6 +102,21 @@ function BlockPattern( {
+								<Tooltip
+									position="top center"
+									text={ __(
+										'Editing this pattern will also update anywhere it is used'
```
This icon is not tabbable and thus won't announce this text for screen readers. Maybe we should merge this into the description above? (And yep probably it's also an issue in the patterns page in the site editor? 😅 )

glendaviesnz replied:
I wonder if it would be more annoying than useful to hit that for every pattern, especially if tabbing through a filtered list of just synced patterns? Do you think we could look at this as a follow up and cover site editor and post inserter together?

kevin940726 replied:
Yes I meant as a follow-up to look into in the future!

kevin940726 commented on packages/block-editor/src/components/block-patterns-paging/index.js:
```diff
@@ -0,0 +1,83 @@
+					<Button
+						variant="tertiary"
+						onClick={ () => changePage( 1 ) }
+						disabled={ currentPage === 1 }
```
I wonder where we should move the focus after navigating to a different page. Should we retain the focus on the prev/next button? Or should we move the focus to the first pattern to match the visual jump? If we decide to retain the focus then we probably want to change this to `aria-disabled` to prevent focus lost when pressing this button for instance. This can all be reviewed by an accessibility expert though, I'm just tossing out some questions 😅 .

glendaviesnz replied:
I think it would be good to look at this as a follow-up combined with the site editor pagination - ideally we should finalise the `pattern` we should use for this and see if we can have a generic pagination component that can be shared between site/post editor.

kevin940726 commented on packages/block-editor/src/components/inserter/hooks/use-patterns-paging.js:
```diff
@@ -0,0 +1,63 @@
+export default function usePatternsPaging(
+	currentCategoryPatterns,
+	currentCategory,
+	scrollContainerClass,
```
Possibly a better API would be accepting a `scrollingContainerRef` to call `scrollToTop` from.

glendaviesnz replied:
This would work in the tab, but in the [patterns explorer ](https://github.com/WordPress/gutenberg/pull/54007/files#diff-6b4f2493a44f524f5f43e18b6a3a655fb59d8306e5d981fafd8363f9b96ded9dR119) the scrollable element is the modal components content area and I couldn't see a way to access the ref from that from within patterns-list component - can you see any way that we could get a ref to that scrollable element?

kevin940726 replied:
Would `wp.dom.getScrollContainer` work?

glendaviesnz replied:
Nice! that works, so I have [switched to using a ref with that method](https://github.com/WordPress/gutenberg/pull/54007/commits/2f53ea25465d4bb9a1911f0aa678fdfbef408257). Thanks.

kevin940726 commented on packages/block-editor/src/components/inserter/hooks/use-patterns-paging.js:
```diff
@@ -0,0 +1,63 @@
+			previousFilter !== currentFilter ) &&
+		currentPage !== 1
+	) {
+		setCurrentPage( 1 );
```
Nit: This works but might be harder to understand or maintain. An easier solution would be resetting the state in event handlers instead. See https://react.dev/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers

glendaviesnz replied:
Yeh, agree that would be better, but currently the select category event handler is two parent components up, so seems like it would be a mission and a bit of refactoring involved to achieve this. I suggest we leave it as is for now and look at this when we come to merge the paging components between site editor and post editor - unless you can see a quick way of hooking this into the `onSelectCategory` event handler that I am overlooking.

dcalhoun commented:
> @dcalhoun, @geriux it would probably be good for someone from mobile to cast their eyes over this one. It merges the `synced patterns/reusable blocks` tab into the `Patterns` tab and adds a filter to limit display to certain types. It looks native generates its own inserter tabs, so I don't think mobile will be affected by this change, but just double-checking. Thanks.

I agree the native variant manages its own tabs. The code does appear to directly impact the native inserter. 

I briefly tested "Synced patterns" in this branch with an iPhone 14 Pro simulator. I did not encounter any regressions. I believe the mobile app is uneffected. 👍🏻

glendaviesnz committed:
Extract filtering logic into a separate method

glendaviesnz committed:
Remove redundant fragment

glendaviesnz committed:
Improve readability of pattern filter

glendaviesnz committed:
Spread paging props instead of destructuring

glendaviesnz committed:
Switch back to all category when changing filters

glendaviesnz committed:
Remove duplicate items from inserter and fix icon alignmen

glendaviesnz committed:
Pattern title layout

glendaviesnz committed:
Fix unit tests

glendaviesnz committed:
Switch icon to center alignment

glendaviesnz committed:
Make the removal of reusable blocks more targeted to keep selector backwards compat

glendaviesnz committed:
Revert "Fix unit tests"

This reverts commit 48dcd35c08db49165d019ab6f67d997071a26fb1.

glendaviesnz commented:
thanks for the review @kevin940726  & @aaronrobertshaw 

> When searching in the inserter you see the synced patterns twice in the results (once as an icon for blocks and a second time in the patterns)

Should be fixed now and display just once under patterns, in inserter tab and also quick inserter search.

> The synced patterns icon looks a bit odd in the search results when it's combined with the pattern name.

Have switched to outline icon and centered it which looks better I think.

> When switching between filters, if the previously selected category isn't in the available categories after filtering, it appears like there are no patterns for the filter. Selecting one of the available categories does correctly show their patterns though.

Should always return to `All` category when switching filters now.

> We probably need to cover the "Show button text labels" preference for the pagination as well

How do you feel about doing this in a follow-up? It looks like the `showIconLabels` preference is currently only used in `edit-post` and `edit-site` so we will need to work out the best way to access that in `block-editor`, which will be a slightly bigger job - unless I am missing something obvious in looking into this.

glendaviesnz committed:
Add show icon labels support to pagination

jameskoster commented:
Nice work on the PR so far, the functionality is very neat!

It probably warrants a more detailed design review, but I wanted to note that combining the author and sync status in a single control makes it trickier to wrap your head around. I'm wondering if the dropdown should focus purely on author with options like:

* All
* My patterns
* Theme patterns
* WooCommerce patterns

Then each flyout could conditionally employ the sync toggle from the main patterns view, roughly like:

<img width="324" alt="Screenshot 2023-09-08 at 13 13 11" src="https://github.com/WordPress/gutenberg/assets/846565/bb8fde65-b03d-4014-b00f-a3d43bf03fa8">

---

The placement of the synced icon still feels a bit strange, and sometimes it reveals an out-of-context tooltip:

<img width="363" alt="Screenshot 2023-09-08 at 13 19 27" src="https://github.com/WordPress/gutenberg/assets/846565/08f670ff-58e4-4451-9b3f-4798ea034ac3">

It might be worth trying no icon, but purple accents on hover.

paaljoachim commented:
Just a general comment here....
This looks really helpful!
Being able to add filtering is a nice way to sort patterns.

Thank you for working on this Glen!

glendaviesnz committed:
Add the show-icons-label to the body element so it also gets applied to modals in the block editor

alexstine commented:
@glendaviesnz After selecting an option from the simple select, the screen reader alert fires several times. I am not sure what causes that to happen but it should be fixed.

Thanks.

glendaviesnz committed:
Switch to using a ref and getScrollContainer to find scroll container

glendaviesnz committed:
remove the synced icon and instead use synced colored border on hover

glendaviesnz committed:
Fix issues with category counts being announced more than once by screen reader

glendaviesnz commented:
> @glendaviesnz After selecting an option from the simple select, the screen reader alert fires several times. I am not sure what causes that to happen but it should be fixed

Thanks @alexstine, I think I have fixed this now.

glendaviesnz committed:
Revert "remove the synced icon and instead use synced colored border on hover"

This reverts commit bf748865941dfcd9f27653276537d08e5f99672c.

glendaviesnz committed:
Put the icon back

glendaviesnz commented:
> It might be worth trying no icon, but purple accents on hover.

Thanks @jameskoster, I have had another look at this today in conjunction with @SaxonF  and for starters we have removed the tooltip and added the title next to the icon to match the design/layout of the user patterns in the site editor pattern library:
<img width="297" alt="Screenshot 2023-09-11 at 2 56 04 PM" src="https://github.com/WordPress/gutenberg/assets/3629020/780012ac-40ae-495c-8578-54b21168cb6b">

I am still working on adapting the filters in line with what you suggested - there are some issues with the sync status toggle not fitting in that narrow panel with longer language translations, so just playing around with options to get around that.

glendaviesnz committed:
split out the sync toggle to pattern list page

jasmussen commented:
[This segmented control feels](https://github.com/WordPress/gutenberg/pull/54007#issuecomment-1711586729) right for the UI, as it's also used in the patterns section itself.

glendaviesnz commented:
> [This segmented control feels](https://github.com/WordPress/gutenberg/pull/54007#issuecomment-1711586729) right for the UI, as it's also used in the patterns section itself.

Thanks @jasmussen the only potential issue is that it just fits into the width of that panel with English, so there may be display issues with longer language strings like German, so I will explore some options for dealing with that.

jasmussen commented:
Indeed, it seems like the segmented control itself should be able to turn into a dropdown if the items do not fit. Should this be separate?

aaronrobertshaw commented:
> Indeed, it seems like the segmented control itself should be able to turn into a dropdown if the items do not fit. Should this be separate?

If that component work could be separated out it would help ensure we land custom pattern categories for WP 6.4. 

The risk then is that the component work cannot be completed before the next GB release, leaving the issue @glendaviesnz [commented](https://github.com/WordPress/gutenberg/pull/54007#issuecomment-1713268979) on. Unless that can be considered a bug?

Do we have any other cases where a segmented control like this doesn't fit width-wise when in non-English languages?

jameskoster commented:
@glendaviesnz @SaxonF should we truncate the titles in this context? 

Could the ToggleGroupControl labels also truncate (currently they wrap)? There's a `tooltip` prop we might use to ensure the labels are always readable.

Possibly a wild idea that I acknowledge won't cater to all languages, but could the patterns column be wider? I always found it a bit backwards that the navigation column (left) is wider the the more visual pattern preview column.

jasmussen commented:
We could also use icons?

jameskoster commented:
Icons are certainly an option, but it might be tricky to come up with ones that communicate the concepts here?

glendaviesnz committed:
Tidy up naming

glendaviesnz committed:
Set height of select filter

glendaviesnz committed:
Reset categories if changing sync filter

glendaviesnz committed:
Reset category when changing sync filter in patterns explorer

glendaviesnz committed:
remove sync filter when searching

glendaviesnz committed:
constrain width of sync toggle in pattern explorer

glendaviesnz committed:
Also set filter height in patterns explorer

glendaviesnz committed:
Switch to select for sync filter temporarily

glendaviesnz committed:
Fix filter width

glendaviesnz committed:
Reset sync filter setting to all each time it is mounted

glendaviesnz committed:
Revert "Reset sync filter setting to all each time it is mounted"

This reverts commit 0ee4b81f03a07c049be104b8d6b32610a7536d1a.

alexstine commented:
@glendaviesnz Still not fixed. 😞 Whenever I select the Patterns tab, this message keeps repeating over and over.

> 10 category buttons displayed.

Thanks.

glendaviesnz commented:
@jameskoster, @jasmussen thanks for the feedback. I played around with truncating the text but couldn't get it to work given the way the way the toggle nests the text in a number of different flex boxes, I am sure it would be possible but is going to be a pretty sub-optimal experience so didn't think it was worth spending a lot of time on.

We have a bit of time pressure to get all of the new pattern category feature PRs merged and well-tested before 20 Sep so I have made the call to switch the sync toggle to a select list temporarily (see the video on the PR description to see how this looks/works). 

This allows us to fully test these changes and get all the category feature PRs merged. We will then quickly circle back with @SaxonF's help to refactor the sync toggle UI in a follow-up PR ahead of the 6.4 beta.

jasmussen commented:
Sounds good to me.

glendaviesnz committed:
Fix issue with all category auto opening when tab first opened

jameskoster commented:
Agree that seems reasonable, it's probably not a high-traffic control so no need to polish too much here.

A couple of other observations:

* The "Pattern source" label feels a bit unnecessary given the proximity to the "Patterns" tab. Could it be visually hidden?
* Should "Sync type" be "Syncing" to match the pattern details panel?
* "Unfiltered" is a bit technical, should it just be "All"?
  * For consistency the first option in the sync dropdown might be "All" instead of "Any".

It will be a bit strange to have an "All" source filter right next to an "All" category, but I think that's a separate issue. It might be worth removing the "All" category and opening that panel automatically when the patterns tab is active.

glendaviesnz committed:
Suggesting wording changes from design review

glendaviesnz committed:
Revert "Fix issue with all category auto opening when tab first opened"

This reverts commit 5951dad7e22d5786bfb035cbbdafb6e0cb63733b.

glendaviesnz committed:
Remove the all category and have all panel open by default

glendaviesnz commented:
@jameskoster thanks for the extra suggestions, I have made all those changes, including getting rid of the `All` category and just opening the preview panel with `All` when the tab is first selected,  and that flow seems to work pretty well. 

It could still do with some refining but I think it is in a much better 'first cut' place now thanks to your suggestions.

glendaviesnz committed:
Remove spoken announcement of category count from hook as currently to much potential for rerender

glendaviesnz committed:
Move filter above category list

glendaviesnz committed:
map to new array in resolver rather than selector to prevent pointless rerenders

glendaviesnz committed:
Revert "Remove spoken announcement of category count from hook as currently to much potential for rerender"

This reverts commit 5fd5edc44fda436934dc8877ea681014281ffcf6.

jameskoster commented:
Thanks for the updates!

> including getting rid of the All category and just opening the preview panel with All when the tab is first selected, and that flow seems to work pretty well.

Agree it works quite well, but there's an issue—you can't get back to viewing the 'All' category in an intuitive way. Let's look into this in a separate follow-up?

---

There's some awkward hover interactions going on – likely due to the addition of the title. I'll see if I can fix that.

jameskoster committed:
Simplify styling

jameskoster commented:
@glendaviesnz one issue I noticed, unsure if it's related to this PR: If you click the "Explore all patterns" button, the modal doesn't include any patterns.

jameskoster commented:
@glendaviesnz a final question: is the pagination UI part of this PR? It looks a little rough. I can adjust that tomorrow if need be.

glendaviesnz committed:
Add the All categories option back so there is an easy way back to listing all

glendaviesnz committed:
Remove onUnmount category reset

glendaviesnz committed:
Fix sync filter reset check

glendaviesnz commented:
> @glendaviesnz one issue I noticed, unsure if it's related to this PR: If you click the "Explore all patterns" button, the modal doesn't include any patterns.

The modal shows patterns for me @jameskoster:

https://github.com/WordPress/gutenberg/assets/3629020/2513b47e-49bb-4ed8-93d6-21cfb51f35fa

Do you see any errors in the console when you load the modal?

glendaviesnz commented:
> Agree it works quite well, but there's an issue—you can't get back to viewing the 'All' category in an intuitive way. Let's look into this in a separate follow-up?

yeh, and it turned out that opening that panel by default was not great for keyboard accessibility, so have added back the All category for now and differentiated the `Alls` as `All pattern types` and `All categories`  - should be opportunity to tweak/refine that in follow up PRs.

glendaviesnz committed:
Add no results message

glendaviesnz commented:
> @glendaviesnz a final question: is the pagination UI part of this PR? It looks a little rough. I can adjust that tomorrow if need be.
 
or if we get this tested and signed off today it will be merged into the [main categories PR](https://github.com/WordPress/gutenberg/pull/53835) and paging could be adjusted there.

glendaviesnz committed:
Reset the category when leaving patterns tab

kevin940726 commented on packages/block-editor/src/components/inserter/block-patterns-explorer/explorer.js:
```diff
@@ -2,33 +2,69 @@
+
+	const previousSyncFilter = usePrevious( patternSyncFilter );
+
+	// If the sync filter changes, we need to select the "All" category to avoid
+	// showing a confusing no results screen.
+	useEffect( () => {
+		if ( patternSyncFilter && patternSyncFilter !== previousSyncFilter ) {
+			setSelectedCategory( initialCategory?.name );
+		}
+	}, [
+		patternSyncFilter,
+		previousSyncFilter,
+		patternSourceFilter,
+		initialCategory?.name,
+	] );
```
Nit: Could we rewrite this without the `useEffect`? Or even better if we could pass `setSelectedCategory` down and call it when the sync filter changes.

glendaviesnz replied:
Changed.

glendaviesnz replied:
yeh it would be nice to do that, but had a look and plumbing through the onCategorySelect  method from modal.js file and using that causes the whole component tree to rerender and the modal gets closed. But would be great to refactor this component tree at some point to allow away for these changes to be handled in the relevant event handlers

kevin940726 commented on packages/block-editor/src/components/inserter/block-patterns-tab.js:
```diff
@@ -205,24 +310,60 @@ function BlockPatternsTabs( {
+	}, [] );
+	const previousSyncFilter = usePrevious( patternSyncFilter );
+
+	// If the sync filter changes, we need to select the "All" category to avoid
+	// showing a confusing no results screen.
+	useEffect( () => {
+		if ( patternSyncFilter && patternSyncFilter !== previousSyncFilter ) {
+			onSelectCategory( allPatternsCategory, patternSourceFilter );
+		}
+	}, [
+		patternSyncFilter,
+		previousSyncFilter,
+		onSelectCategory,
+		patternSourceFilter,
+	] );
```
Nit: Same here. Could we directly set the state in the render phase to avoid an extra render? Never mind if it's not possible though!

glendaviesnz committed:
make sure pattern type filters are applied to search results

kevin940726 commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -220,11 +220,17 @@ function Layout() {
 		[ entitiesSavedStatesCallback ]
 	);
 
+	// We need to add the show-icon-labels class to the body element so it is applied to modals.
+	if ( showIconLabels ) {
+		document.body.classList.add( 'show-icon-labels' );
+	} else {
+		document.body.classList.remove( 'show-icon-labels' );
+	}
```
Since it's a side effect, should we put this in an `useEffect`?

glendaviesnz replied:
yeh, have moved to a useEffect

kevin940726 reviewed (approved):
Looking pretty good! Just a couple of minor suggestions and none are blockers! 💯

glendaviesnz commented:
This was merged with some failing tests, but is going into a feature branch which needs rebasing to get the failing tests resolved, so failing tests will be sorted on that branch prior to merging to trunk.
